### PR TITLE
Augment NVTX annotations attached to HLO modules and thunks

### PR DESCRIPTION
**Pull request by olupton**
This builds on the work in #7106 to make the NVTX annotations shown in the NSight Systems profiler more helpful.

Now the string representation of the HLO, and the Python stack trace of the traced model code, are shown in these annotations. For example, for a cuBLAS kernel:

```
Thunk:#name=while/body/transpose[permutation=(1, 0)],hlo_op=custom-call.49#
Begins: 13.3989s
Ends: 13.3989s (+10.209 μs)
Thread: 42903
XlaKernel
Source locations: 
- /opt/jax/examples/mnist_vae.py:143[<module>]
-- /opt/jax/examples/mnist_vae.py:127[run_epoch]
--- /opt/jax/examples/mnist_vae.py:125[body_fun]
---- /opt/jax/examples/mnist_vae.py:124[<lambda>]
----- /opt/jax/examples/mnist_vae.py:122[body_fun]

HLO: custom-call.49 = (f32[512,20]{1,0}, s8[68096]{0}) custom-call(fusion.46, wrapped_concatenate.28), custom_call_target="__cublas$gemm", metadata={op_name="jit(run_epoch)/jit(main)/while/body/transpose[permutation=(1, 0)]" source_file="/opt/jax/examples/mnist_vae.py" source_line=122}, backend_config={"operation_queue_id":"0","wait_on_operation_queues":[],"gemm_backend_config":{"alpha_real":1,"alpha_imag":0,"beta":0,"dot_dimension_numbers":{"lhs_contracting_dimensions":["0"],"rhs_contracting_dimensions":["0"],"lhs_batch_dimensions":[],"rhs_batch_dimensions":[]},"precision_config":{"operand_precision":["DEFAULT","DEFAULT"]},"epilogue":"DEFAULT","lhs_stride":"16384","rhs_stride":"640","grad_x":false,"grad_y":false}}
```
or with a fusion:
```
Source locations: 
- /opt/jax/examples/mnist_vae.py:143[<module>]
-- /opt/jax/examples/mnist_vae.py:127[run_epoch]
--- /opt/jax/examples/mnist_vae.py:126[body_fun]

HLO: fusion.133 = f32[512,512]{1,0} fusion(get-tuple-element.681), kind=kLoop, calls=fused_computation.7, metadata={op_name="jit(run_epoch)/jit(main)/while/body/mul" source_file="/opt/jax/examples/mnist_vae.py" source_line=126 deduplicated_name="fusion.7"}
Called HLO: 
fused_computation.7 {
  param_0.9 = f32[512,512]{1,0} parameter(0)
  constant_378 = f32[] constant(0.9)
  broadcast.575 = f32[512,512]{1,0} broadcast(constant_378), dimensions={}
  ROOT multiply.250 = f32[512,512]{1,0} multiply(param_0.9, broadcast.575), metadata={op_name="jit(run_epoch)/jit(main)/while/body/mul" source_file="/opt/jax/examples/mnist_vae.py" source_line=126}
}
```

### CODE REVIEW

File: xla/service/gpu/thunk.h
Comment (ezhulenev) [inline xla/service/gpu/thunk.h:330]: std::string_view

File: xla/service/gpu/runtime3/sequential_thunk.cc
Comment (ezhulenev) [inline xla/service/gpu/runtime3/sequential_thunk.cc:58]: `annotations = Get...`

**Review by ezhulenev — APPROVED** [review]

**Comment by ddunl** [comment]
Copybara is struggling with this one, so I'm going to land just the changes that add nvtx as a dependency with the associated Copybara changes, then this should be good to import (but will require rebase on your side I think). Sorry for the delay!

**Comment by olupton** [comment]
> Copybara is struggling with this one, so I'm going to land just the changes that add nvtx as a dependency with the associated Copybara changes, then this should be good to import (but will require rebase on your side I think). Sorry for the delay!

Thanks! Rebased it now.

**Review by ddunl — APPROVED** [review]
Thanks!!

**Comment by olupton** [comment]
I rebased again and pushed a fix for non-GPU builds (430faf37b1a6cdd8726b1463bc1c6a64638cebb2) that made one of the CI jobs green. Hopefully this is ready to land now.

**Review by ddunl — APPROVED** [review]
Thanks!!

**Comment by olupton** [comment]
@ddunl looks like cl/601810646 failed, please let me know if you need anything from my side.

**Comment by ddunl** [comment]
This is another copybara fix, I'll try to send a change to fix today. Sorry for the delay!!

**Comment by ddunl** [comment]
Sorry about the delay again, will get to this today/tomorrow

**Comment by ddunl** [comment]
Sorry for delay again, seems like I need to make an internal change to be able to use `"nvtx3/nvToolsExtPayload.h"` internally

**Comment by ddunl** [comment]
I'm trying to figure out exactly what needs to happen for this header to be used internally, but in the meantime it would be great if you could rebase

**Comment by olupton** [comment]
> I'm trying to figure out exactly what needs to happen for this header to be used internally, but in the meantime it would be great if you could rebase

Done :+1:

**Review by ddunl — APPROVED** [review]
Thanks! Still on me to figure out this header today.

**Comment by chsigg** [comment]
Olli, could you explain what the plans are wrt exposing the payload API in NVTX? Is this going to ship with CUDA 12.4? Will https://github.com/NVIDIA/NVTX be updated soon? Thanks!

**Comment by olupton** [comment]
> Olli, could you explain what the plans are wrt exposing the payload API in NVTX? Is this going to ship with CUDA 12.4? Will https://github.com/NVIDIA/NVTX be updated soon? Thanks!

I don't have a precise timeline for this, but they should become available through those channels in due course.
In the meantime, https://github.com/openxla/xla/blob/bbb68a483faffdd23b4221f84e873532a941cdbf/third_party/tsl/workspace2.bzl#L434-L441 should already work? That landed in ddffe4d0bf8a9d075578b26150546c9229f2aba8.

I will push another rebase shortly.

**Comment by olupton** [comment]
@ezhulenev I pushed a rebase on top of your reorganisation of runtime/, runtime3/ and so on. I hope this follows your intent.

Also note that I added annotations in a handful more command_buffer-related places. This helps Nsight Systems to associate the NVTX metadata with CUDA graph nodes when they execute on the device.

**Comment by chsigg** [comment]
> should already work? That landed in [ddffe4d](https://github.com/openxla/xla/commit/ddffe4d0bf8a9d075578b26150546c9229f2aba8).

That's our only option right now to land this PR, but given how tricky it has been so far it seems not ideal.

**Comment by olupton** [comment]
> > should already work? That landed in [ddffe4d](https://github.com/openxla/xla/commit/ddffe4d0bf8a9d075578b26150546c9229f2aba8).
> 
> That's our only option right now to land this PR, but given how tricky it has been so far it seems not ideal.

We don't have a precise timeline, but we anticipate that it will appear through https://github.com/NVIDIA/NVTX first - so it should "just" be an update of the `urls`, `sha256` and `strip_prefix` in https://github.com/openxla/xla/blob/bbb68a483faffdd23b4221f84e873532a941cdbf/third_party/tsl/workspace2.bzl#L434-L441

Regarding the GitHub CI failure (https://storage.cloud.google.com/xla-public/prod/tensorflow/xla/linux/gpu/build_gpu/35446/20240208-041546/index.html), it looks like this is failing on other PRs too so I have not investigated further.

**Comment by olupton** [comment]
I rebased this again, mainly due to conflict with ab7777bb08e0d29aa12b3c47d37df4b6b3369bc3 but also some light clashes with the `runtime3` -> `runtime` naming.

Please let me know if there's anything we could do to help unblock this. I don't fully understand what the internal issue is, given that it all seems to work fine in OSS.

**Comment by ddunl** [comment]
I think we have a resolution internally, so I'll submit my CL early tomorrow and then we should be good to go. Sorry for the delay here

**Comment by ddunl** [comment]
Alright, should just need to rebase this PR one more time (sorry) and we should be good to go. Sorry for the delay again.

**Comment by olupton** [comment]
I rebased it again. 🤞 that's the last time 🙂

**Review by akuegel — APPROVED** [review]

**Review by akuegel — CHANGES_REQUESTED** [review]
It looks like the rebase didn't update the code according to https://github.com/openxla/xla/commit/ab7777bb08e0d29aa12b3c47d37df4b6b3369bc3
We get compile errors, for example from ir_emitter_context.

**Review by akuegel — APPROVED** [review]
Approving again, the problem is on our side, we didn't fetch one of the files from this PR (nvtx_utils.h) @ddunl FYI

**Comment by nouiz** [comment]
It seems the pull didn't work again.

**Comment by akuegel** [comment]
> It seems the pull didn't work again.

As far as I can tell, it worked fine. What you can see here are two different PRs, one for Openxla, one for TSL. This is expected.

If you mean this has not been merged yet, then this is due to not getting approval internally. Several review comments have been added which need to be addressed, and I was out of office yesterday so I couldn't work on it yet. I will do the required changes myself to speedup the process.

**Comment by ddunl** [comment]
I think we are merging the PR as is, but it would be really great if https://github.com/NVIDIA/NVTX could be updated :), it would allow us to have a much nicer situation wrt headers internally

**Pull request closed without merging**
